### PR TITLE
Adding Sanitizer interface and implementation

### LINKED ISSUE #8: Add Metrics Sanitizer
Add the tally-java equivalent of https://github.com/uber-go/tally/pull/50 

Note: use `Sanitize`, not `Sanitise`

### PR DESCRIPTION
**Pull request by ravirajj**
- Sanitizes metric names and tag keys and values.
- Does nothing by default.
- The `ScopeBulider` can be configured with the standard `M3Sanitizer`.
- Fixes #8

### CODE REVIEW

File: core/src/main/java/com/uber/m3/tally/sanitizers/Sanitize.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.uber.m3.tally.sanitizers;
+
+/**
+ * Sanitize returns a sanitized version of the input string value.
+ */
+public interface Sanitize {
```
Comment (inetchitailo) [inline core/src/main/java/com/uber/m3/tally/sanitizers/Sanitize.java:26]: Please try for a different name. `Sanitize` next to `Sanitizer` confused me thoroughly

File: core/src/main/java/com/uber/m3/tally/sanitizers/ValidCharacters.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    Sanitize sanitize(char repChar) {
+        return value -> {
+            StringBuilder buffer = null;
+
+            for (int i = 0; i < value.length(); i++) {
+                char ch = value.charAt(i);
+
+                // first check if the provided character is valid
+                boolean validCurr =
+                    ranges.stream().anyMatch(range -> ch >= range.low() && ch <= range.high())
```
Comment (inetchitailo) [inline core/src/main/java/com/uber/m3/tally/sanitizers/ValidCharacters.java:82]: is this operation performance-sensitive? If so, this looks very expensive

File: core/src/main/java/com/uber/m3/tally/sanitizers/Sanitize.java (hunk 1 above)
Comment (ravirajj) [reply core/src/main/java/com/uber/m3/tally/sanitizers/Sanitize.java:26]: The `Sanitizer` interface matches the `uber-go/tally` [Sanitizer](https://github.com/uber-go/tally/blob/master/sanitize.go#L77) interface.

The `Sanitize` interface is meant to match the `uber-go/tally` [SanitizeFn](https://github.com/uber-go/tally/blob/master/sanitize.go#L57) interface.  Name it `SanitizeFn` too?

File: core/src/main/java/com/uber/m3/tally/sanitizers/ValidCharacters.java (hunk 2 above)
Comment (ravirajj) [reply core/src/main/java/com/uber/m3/tally/sanitizers/ValidCharacters.java:82]: It matches the `uber-go/tally` [sanitzeFn()](https://github.com/uber-go/tally/blob/master/sanitize.go#L127).  Like the go version, it optimizes allocations by avoiding copying if the all characters are valid.

**Comment by longquanzheng** [comment]
Any progress on this PR? @ravirajj @inetchitailo ?

**Comment by longquanzheng** [comment]
@SokolAndrey @andrewmains12 @alexeykudinkin can you help take a look? I am needing this in my project. Thanks so much

File: core/src/main/java/com/uber/m3/tally/ScopeImpl.java [hunk 3]
```diff
@@ -313,6 +320,14 @@ public Snapshot snapshot() {
         return snap;
     }
 
+    private ImmutableMap<String, String> copyAndSanitizeMap(Map<String, String> tags) {
```
Comment (alexeykudinkin) [inline core/src/main/java/com/uber/m3/tally/ScopeImpl.java:323]: Let's avoid copying if there are no changes to the map

File: core/src/main/java/com/uber/m3/tally/sanitizers/Sanitizer.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * SanitizerImpl sanitizes the provided input based on the function executed.
+ */
+public interface Sanitizer {
+
+    /**
+     * Name sanitizes the provided 'name' string.
+     * @param name the name string
+     * @return the sanitized name
+     */
+    String name(String name);
```
Comment (alexeykudinkin) [inline core/src/main/java/com/uber/m3/tally/sanitizers/Sanitizer.java:33]: Please follow the naming convention -- method name should bear a verb in it

File: core/src/main/java/com/uber/m3/tally/sanitizers/Sanitize.java (hunk 1 above)
Comment (alexeykudinkin) [reply core/src/main/java/com/uber/m3/tally/sanitizers/Sanitize.java:26]: Please elaborate on what this interface is needed for

**Review by alexeykudinkin — CHANGES_REQUESTED** [review]

**Comment by longquanzheng** [comment]
@alexeykudinkin @ravirajj Hello, thanks so much for this. 
Would you mind if I fork this branch and open another PR to address the comments? 

Thanks

**Comment by alexeykudinkin** [comment]
@longquanzheng works for me

**Pull request closed without merging**